File: Lists/LinkedList.cpp
#include "stdafx.h"

#include "LinkedList.h"
#include "Node.h"

LinkedList::~LinkedList()
{
	Node *node = head;
	while (node)
	{
		Node *t = node;
		node = node->GetNext();
		delete t;
	}
}
void LinkedList::Print()
{
	Node *node = head;
	while (node)
	{
		printf("%c", node->GetData());
		node = node->GetNext();
	}
	printf("\n");
}

void LinkedList::Add(Node *node)
{
	if (NULL == head)
	{
		head = node;
		tail = node;
	}
	else
	{
		tail->SetNext(node);
		tail = node;
	}
}

void LinkedList::RemoveDuplicateNodes()
{
	Node *node;

	node = head;

	while (node)
	{
		Node *run = node;
		while (run->GetNext())
		{
			if (node->GetData() == run->GetNext()->GetData())
			{
				Node *del = run->GetNext();
				run->SetNext(del->GetNext());
				delete del;
			}
			run = run->GetNext();
		}
		node = node->GetNext();
	}
}

File: src/main/java/cracking/linkedlist/LinkedList.java
package cracking.linkedlist;

public class LinkedList {
	private Node head;
	private Node tail;
	
	LinkedList() {
		head = null;
		tail = null;
	}
	
	void AddNode(Node node) {
		if (null == head) {
			head = node;
			tail = node;
		}
		else {
			tail.next = node;
			tail = node;			
		}
	}
	public void Add(String string) {
		Node node = new Node(string);
		
		if (null == head) {
			head = node;
			tail = node;
		}
		else {
			tail.next = node;
			tail = node;			
		}
	}
	
	Node FindBeginning() {
		Node slow = head;
		Node fast = head;
		
		while (fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			if (slow == fast) {
				break;
			}
		}
		
		if (fast == null || fast.next == null) {
			return fast;
		}
		
		slow = head;
		while (slow != fast) {
			slow = slow.next;
			fast = fast.next;
		}
		return fast;
	}
}
